### PR TITLE
Improve stdin parsing

### PR DESCRIPTION
**Pull request by chriscool**
This should fix issue #1141 (ipfs cat "multihash too short" error when using stdin).

### CODE REVIEW

**Comment by chriscool** [comment]
Some tests failed with:
```
No output has been received in the last 10 minutes, this potentially indicates a stalled build or something wrong with the build itself.
```
and one with:
```
FAIL	github.com/ipfs/go-ipfs/routing/dht
```

File: commands/cli/parse.go [hunk 1]
```diff
@@ -255,13 +255,17 @@ func parseArgs(inputs []string, stdin *os.File, argDefs []cmds.Argument, recursi
...
-					return nil, nil, err
+				if len(inputs) > 0 {
+					// don't use stdin if we have inputs
+					stdin = nil
+				} else {
+					// if we have a stdin, read it in and use the data as a string value
+					stringArgs, stdin, err = appendStdinAsString(stringArgs, stdin)
+					if err != nil {
+						return nil, nil, err
+					}
```
Comment (jbenet) [inline commands/cli/parse.go]: im not sure this works with all commands-- do we have a case where we use both arguments AND stdin? maybe some of the `ipfs object put` type things? (although those commands are not very intuitive right now, so perhaps they can change)

Comment (jbenet) [inline commands/cli/parse.go]: in general, this part of the cmdslib is a bit of a mess. ideas to clean it up would be great.

**Comment by jbenet** [comment]
@whyrusleeping PTAL?

**Comment by chriscool** [comment]
@jbenet @whyrusleeping I just added more improvements, tests and a fix for issue #1196 (Can't launch a command line process from Qt).
`ipfs object put` is a command like "stdinenablednotvariadic" in https://github.com/ipfs/go-ipfs/commit/275ec7c2a00c2cbd9d5b1bc00a6e097236f29851 so there are now tests for commands like it too.
And there are now tests for commands than can take both an argument and stdin too (see tests for "stdinenabled2args" and "stdinenablednotvariadic2args").

**Comment by chriscool** [comment]
The errors in Travis are the same "No output has been received in the last 10 minutes..."

**Comment by whyrusleeping** [comment]
@chriscool this LGTM, those travis failures are strange, but not your fault.

**Comment by jbenet** [comment]
rerunning---

We are using the KVM setup-- which is faster-- though is still beta.

**Comment by whyrusleeping** [comment]
we should either set the go test timeout to 9 minutes, or the travis test timeout to 11 minutes, its hard to tell whether things broke because travis, or if things broke because one of our tests hung

**Comment by jbenet** [comment]
@whyrusleeping sgtm

**Comment by chriscool** [comment]
There are still the same kind of failures in some go tests:

```
*** Test killed with quit: ran too long (10m0s).
FAIL	github.com/ipfs/go-ipfs/merkledag	600.012s
```

and

```
--- FAIL: TestBootstrap (0.89s)
	dht_test.go:314: connecting 30 dhts in a ring
	dht_test.go:77: dial attempt failed: misdial to <peer.ID Y91TvF> through /ip4/127.0.0.1/tcp/41134 (got <peer.ID dTyTNY>)
18:47:07.334 CRITI   boguskey: TestBogusPrivateKey.Sign -- this better be a test! key.go:54
FAIL
FAIL	github.com/ipfs/go-ipfs/routing/dht	3.503s
```

**Pull request merged**